File: frontend/src/api/portfolio.ts
import { apiClient } from './client'

/** One holding in the portfolio being tested. */
export interface PortfolioHolding {
  symbol: string
  exchange: string
  /** Percentage or fraction — only the ratio between holdings matters. */
  weight: number
}

export type RebalanceRule = 'never' | 'monthly' | 'quarterly' | 'yearly'
/** 'db' reads the local Historify store; 'api' asks the broker. */
export type PriceSource = 'db' | 'api'

export interface BacktestRequest {
  apikey: string
  holdings: PortfolioHolding[]
  start_date: string
  end_date: string
  benchmark?: string | null
  benchmark_exchange?: string
  rebalance?: RebalanceRule
  drift_band?: number
  cost_model?: 'indian_equity' | 'flat_bps'
  brokerage_pct?: number
  cost_bps?: number
  cost_exchange?: string
  /** Per-charge overrides, keyed by charge id. Rates are fractions. */
  charges?: Record<string, { rate?: number; flat?: number; cap?: number }>
  /** Consumption tax on the taxable charges, as a fraction. */
  gst_rate?: number
  slippage?: number
  initial_capital?: number
  risk_free_rate?: number
  source?: PriceSource
}

export interface CurvePoint {
  date: string
  value: number
}

/**
 * Per-holding P&L. `contribution_pct` is this holding's share of the portfolio
 * return, so the column sums to the total — which is what makes it an
 * attribution rather than a list of individual performances. It differs from
 * `symbol_return` whenever the weight is not 100%.
 */
export interface PortfolioItem {
  symbol: string
  weight_target: number
  weight_final: number
  invested: number
  price_pnl: number
  costs: number
  net_pnl: number
  contribution_pct: number
  symbol_return: number
}

/** Anything the engine could not compute honestly comes back null, not 0. */
export interface PortfolioMetrics {
  cagr: number | null
  volatility: number | null
  sharpe: number | null
  sortino: number | null
  calmar: number | null
  max_drawdown: number | null
  win_rate: number | null
  best_day: number | null
  worst_day: number | null
  value_at_risk: number | null
  cvar: number | null
  ulcer_index: number | null
  recovery_factor: number | null
  tail_ratio: number | null
  skew: number | null
  kurtosis: number | null
  alpha?: number | null
  beta?: number | null
  information_ratio?: number | null
  benchmark_cagr?: number | null
  excess_cagr?: number | null
  up_capture?: number | null
  down_capture?: number | null
}

/**
 * One scored dimension of portfolio health, with its working attached so the
 * grade can be argued with rather than merely believed.
 */
export interface HealthPillar {
  key: string
  label: string
  /** 0-100, or null when the pillar could not be measured. */
  score: number | null
  /** Its share of the total as designed. */
  weight: number
  /** Its share after renormalising around any unmeasured pillar. */
  effective_weight: number
  formula: string
  inputs: Record<string, unknown>
  comment: string
}

export interface PortfolioHealth {
  score: number | null
  grade: string | null
  pillars: HealthPillar[]
  /** Pillars dropped for want of data, rather than scored zero. */
  unmeasured: string[]
}

/** Series-level data: what a table of headline numbers cannot convey. */
export interface PortfolioSeries {
  drawdown: CurvePoint[]
  drawdown_episodes?: {
    start: string
    valley: string
    end: string
    days: number
    depth: number | null
  }[]
  rolling?: {
    window: number
    sharpe: CurvePoint[]
    volatility: CurvePoint[]
  }
  monthly_returns?: {
    years: string[]
    columns: string[]
    values: (number | null)[][]
  }
  /** Best/worst single month, typical up/down month, win rate by month and quarter. */
  seasonality?: {
    best_month: number | null
    worst_month: number | null
    avg_up_month: number | null
    avg_down_month: number | null
    win_months: number | null
    win_quarters: number | null
  }
  weekly_returns: CurvePoint[]
  /** Weekly returns as years x ISO weeks, for the heatmap. */
  weekly_grid?: {
    years: string[]
    weeks: number[]
    values: (number | null)[][]
  }
  /** Spread of returns at each holding period. */
  return_quantiles?: {
    period: string
    count: number
    min: number | null
    q1: number | null
    median: number | null
    q3: number | null
    max: number | null
    mean: number | null
    negative_share: number | null
  }[]
  /** Calendar-year returns, against the benchmark where there is one. */
  yearly_returns?: {
    year: string
    portfolio: number | null
    benchmark?: number | null
    difference?: number | null
    /** Straight from openstatz, so the verdict matches its tearsheet. */
    won?: boolean
    /** Omitted when the benchmark fell — a negative denominator is unreadable. */
    multiplier?: number | null
  }[]
}

/**
 * What the trading actually cost, itemised the way a contract note is.
 * STT dominates and brokerage is usually zero — which is why "zero brokerage"
 * does not make rebalancing free.
 */
export interface PortfolioCosts {
  model: string
  brokerage?: number
  stt?: number
  exchange_txn?: number
  sebi?: number
  gst?: number
  stamp_duty?: number
  slippage?: number
  total: number
  drag: number
  turnover: number
}

/** How the portfolio behaved across named historical stress windows. */
export interface PortfolioCrisis {
  periods: {
    key: string
    label: string
    note?: string
    start: string
    end: string
    scope?: 'india' | 'global'
    /** Calendar days and trading sessions the window actually covered. */
    days?: number
    sessions?: number
    portfolio: number | null
    benchmark: number | null
    excess: number | null
    coverage?: number
    partial?: boolean
  }[]
  summary: {
    count: number
    average_return: number
    hit_rate: number | null
    worst: number
    best: number
    by_scope?: Record<
      string,
      { count: number; average_return: number; hit_rate: number | null }
    >
  } | null
}

/** Weights over time, for the allocation chart. No cash band: long-only. */
export interface PortfolioAllocation {
  dates: string[]
  symbols: string[]
  series: Record<string, number[]>
  average: Record<string, number>
}

interface Band {
  p05: number
  p25: number
  median: number
  p75: number
  p95: number
}

/** The same allocation re-run over rolling sub-windows. */
export interface WalkForward {
  windows: {
    start: string
    end: string
    total_return: number
    cagr: number
    max_drawdown: number
    sessions: number
  }[]
  summary: {
    count: number
    window_years: number
    step_years: number
    best: number
    worst: number
    median: number
    positive_share: number
    spread: number
  } | null
  note: string | null
}

/** Block-bootstrap resampling: how much of the outcome was sequence luck. */
export interface MonteCarlo {
  paths: number
  block?: number
  seed?: number
  total_return?: Band
  cagr?: Band
  max_drawdown?: Band
  probability_of_loss?: number
  worst_drawdown_seen?: number
  note: string | null
}

/** What the portfolio is made of, from data that actually exists. */
export interface PortfolioStructure {
  instrument_classes: Record<string, number>
  instrument_class_basis: string
  clusters: { id: number; members: string[]; weight: number; independent: boolean }[]
  threshold: number
  effective_bets: number
  largest_cluster_weight: number
  sector_note: string
}

/** The same allocation under every rebalancing rule. */
export interface RebalancingSweep {
  variants: {
    label: string
    rule: string
    drift_band: number
    total_return: number
    cagr: number
    volatility: number
    sharpe: number
    sortino: number
    max_drawdown: number
    calmar: number
    cost_drag: number
    turnover: number
    rebalances: number
  }[]
  curves: Record<string, CurvePoint[]>
  best_by_sharpe: string | null
  best_by_return: string | null
  sessions: number
  start: string
  end: string
}

/** Where the out- or under-performance came from. */
export interface PortfolioAttribution {
  available: boolean
  reason?: string
  portfolio_return?: number
  equal_weight_return?: number
  benchmark_return?: number
  excess_return?: number
  selection_effect?: number
  allocation_effect?: number
  cost_effect?: number
  holdings?: {
    symbol: string
    weight: number
    return: number
    vs_benchmark: number
    contribution: number
  }[]
  holdings_reason?: string | null
  method?: string
}

export interface BacktestResponse {
  status: 'success' | 'error'
  message?: string
  meta: {
    symbols: string[]
    target_weights: Record<string, number>
    rule: RebalanceRule
    drift_band: number
    cost_bps: number
    slippage: number
    initial_capital: number
    sessions: number
    source: PriceSource
    start: string
    end: string
    benchmark: string | null
    risk_free_rate: number
    /** 'price' — broker history excludes dividends, so returns understate. */
    total_return_basis: string
    /** symbol -> [[date, return]] where a move looked like an unadjusted split. */
    data_warnings: Record<string, [string, number][]>
  }
  equity: CurvePoint[]
  benchmark_equity: CurvePoint[]
  metrics: PortfolioMetrics
  items: PortfolioItem[]
  correlation: {
    symbols: string[]
    /** Square matrix; null where two holdings barely overlap. */
    matrix: (number | null)[][]
    average_pairwise: number | null
  }
  diversification: {
    hhi: number
    effective_holdings: number
    largest_weight: number
    holdings: number
    diversification_ratio: number | null
  }
  walk_forward: WalkForward
  monte_carlo: MonteCarlo
  rebalancing_sweep: RebalancingSweep
  attribution: PortfolioAttribution
  /** Trailing return per holding, close to close. */
  asset_returns: {
    symbol: string
    last: number
    '1W': number | null
    '1M': number | null
    '3M': number | null
    '1Y': number | null
    '3Y': number | null
    '5Y': number | null
  }[]
  structure: PortfolioStructure
  allocation: PortfolioAllocation
  crisis: PortfolioCrisis
  costs: PortfolioCosts
  series: PortfolioSeries
  health: PortfolioHealth
  rebalancing: {
    rule: RebalanceRule
    drift_band: number
    count: number
    cost_drag: number
    turnover_total: number
    dates: string[]
  }
}

/**
 * Run one backtest. A single call returns every tab's data because they are
 * all views of one simulation — fetching per tab would re-run it and risk two
 * tabs disagreeing about the same portfolio.
 */
export async function runPortfolioBacktest(
  req: BacktestRequest
): Promise<BacktestResponse> {
  const { data } = await apiClient.post<BacktestResponse>('/portfolio/backtest', req)
  return data
}

export interface Benchmark {
  symbol: string
  exchange: string
  name: string | null
}

/**
 * Index symbols usable as a benchmark, read from the instrument master.
 *
 * Not hardcoded: which indices exist depends on which broker's master
 * contract was downloaded, and brokers differ on this (GLOBAL_INDEX rows
 * only appear for a broker that supports them).
 */
export async function listBenchmarks(apiKey: string): Promise<Benchmark[]> {
  const { data } = await apiClient.get<{ data: Benchmark[] }>('/portfolio/benchmarks', {
    params: { apikey: apiKey },
  })
  return data.data
}

/**
 * Download the full openstatz tearsheet as a self-contained HTML file.
 *
 * Returned as a blob rather than rendered in-app: it is a report a user keeps
 * and shares, roughly a megabyte with every chart embedded.
 */
export async function downloadTearsheet(req: BacktestRequest): Promise<void> {
  const { data } = await apiClient.post('/portfolio/tearsheet', req, {
    responseType: 'blob',
  })
  const url = URL.createObjectURL(new Blob([data], { type: 'text/html' }))
  const link = document.createElement('a')
  link.href = url
  link.download = 'portfolio-tearsheet.html'
  document.body.appendChild(link)
  link.click()
  link.remove()
  // Revoking immediately can cancel the download in some browsers; a tick is
  // enough for it to have started.
  setTimeout(() => URL.revokeObjectURL(url), 1000)
}

export interface HoldingRow {
  symbol: string
  exchange: string
  quantity: number
  average_price: number
  last_price: number
  invested: number
  current: number
  pnl: number
  /** Null when the broker reports no average price for this holding. */
  pnl_pct: number | null
  weight: number
  product: string
}

export interface HoldingsAnalysis {
  status: 'success' | 'error'
  message?: string
  summary: {
    holdings: HoldingRow[]
    weights: Record<string, number>
    /** Null when the feed carries no average price at all. */
    invested: number | null
    current: number
    pnl: number
    pnl_pct: number | null
    has_cost_basis?: boolean
    count: number
  }
  /** The full backtest report over the current weights, or null if unavailable. */
  analysis: BacktestResponse | null
  analysis_error: string | null
  /** Held, but on an exchange the backtester cannot price. */
  skipped: string[]
  meta: {
    lookback_days: number
    start: string
    end: string
    /** Says plainly that this is today's holdings run over past prices. */
    basis: string
  }
}

/** Analyse the portfolio actually held at the broker. */
export async function analyseHoldings(req: {
  apikey: string
  lookback_days?: number
  benchmark?: string | null
  risk_free_rate?: number
  source?: PriceSource
}): Promise<HoldingsAnalysis> {
  const { data } = await apiClient.post<HoldingsAnalysis>('/portfolio/holdings', req)
  return data
}

File: restx_api/portfolio.py
"""
Portfolio Backtester API.

One POST returns everything the analysis tabs render, because they are all
views of a single simulation -- splitting them across endpoints would mean
re-running the backtest per tab and risking two tabs disagreeing about the
same portfolio.
"""

import os

from flask import jsonify, make_response, request
from flask_restx import Namespace, Resource
from marshmallow import Schema, ValidationError, fields, validate

from database.auth_db import get_auth_token_broker, verify_api_key
from limiter import limiter
from portfolio.data import BENCHMARK_EXCHANGES
from services.portfolio_service import (
    MAX_SYMBOLS,
    analyse_live_holdings,
    generate_tearsheet,
    list_benchmarks,
    run_portfolio_backtest,
)
from utils.logging import get_logger

# Heavier than a quote: a run loads full history for every holding and
# simulates it, so the ceiling is lower than the shared data-API limit.
API_RATE_LIMIT = os.getenv("PORTFOLIO_API_RATE_LIMIT", "10 per minute")
api = Namespace("portfolio", description="Portfolio Backtester API")

logger = get_logger(__name__)

# Cash equity and ETFs only, matching the engine's own allow-list.
PORTFOLIO_EXCHANGES = ["NSE", "BSE"]


class HoldingSchema(Schema):
    symbol = fields.Str(required=True, validate=validate.Length(min=1, max=64))
    exchange = fields.Str(load_default="NSE", validate=validate.OneOf(PORTFOLIO_EXCHANGES))
    # Percentages or fractions both work; only the ratio between them matters.
    weight = fields.Float(required=True, validate=validate.Range(min=0))


class PortfolioBacktestSchema(Schema):
    apikey = fields.Str(required=True, validate=validate.Length(min=1, max=256))
    holdings = fields.List(
        fields.Nested(HoldingSchema),
        required=True,
        validate=validate.Length(min=1, max=MAX_SYMBOLS),
    )
    start_date = fields.Str(required=True)
    end_date = fields.Str(required=True)

    benchmark = fields.Str(load_default=None, allow_none=True)
    # Indices, not the cash exchanges: a benchmark cannot be held, and alpha
    # or beta measured against a single stock would not mean anything.
    benchmark_exchange = fields.Str(
        load_default="NSE_INDEX", validate=validate.OneOf(BENCHMARK_EXCHANGES)
    )
    rebalance = fields.Str(
        load_default="never",
        validate=validate.OneOf(["never", "monthly", "quarterly", "yearly"]),
    )
    # Percentage points of drift that force a rebalance regardless of the
    # calendar. Capped below 1.0 because a band of 100% can never trigger.
    drift_band = fields.Float(load_default=0.0, validate=validate.Range(min=0, max=0.99))
    # The itemised Indian delivery-equity schedule by default; flat bps is
    # kept only so a user can compare against a theoretical rate.
    cost_model = fields.Str(
        load_default="indian_equity",
        validate=validate.OneOf(["indian_equity", "flat_bps"]),
    )
    brokerage_pct = fields.Float(load_default=0.0, validate=validate.Range(min=0, max=0.05))
    # Which exchange's transaction charge applies. Explicit rather than taken
    # from the first holding, since a mixed NSE/BSE book has no single answer.
    cost_exchange = fields.Str(
        load_default="NSE", validate=validate.OneOf(PORTFOLIO_EXCHANGES)
    )
    # Per-charge overrides, so every rate is the caller's to set. Statutory
    # rates change with the budget and differ by market; baking them into the
    # server would mean a release every time one moves.
    charges = fields.Dict(
        keys=fields.Str(),
        values=fields.Dict(
            keys=fields.Str(),
            values=fields.Float(
                allow_none=True,
                validate=validate.Range(min=0),
            ),
        ),
        load_default=dict,
    )
    gst_rate = fields.Float(load_default=None, allow_none=True,
                            validate=validate.Range(min=0, max=1))
    cost_bps = fields.Float(load_default=0.0, validate=validate.Range(min=0, max=1000))
    slippage = fields.Float(load_default=0.0, validate=validate.Range(min=0, max=0.1))
    initial_capital = fields.Float(load_default=100000.0, validate=validate.Range(min=1))
    risk_free_rate = fields.Float(load_default=0.0, validate=validate.Range(min=0, max=0.5))
    # 'db' reads the local Historify store: deterministic and rate-limit free,
    # which is what a long multi-symbol run wants. 'api' asks the broker.
    source = fields.Str(load_default="db", validate=validate.OneOf(["db", "api"]))


backtest_schema = PortfolioBacktestSchema()


@api.route("/benchmarks", strict_slashes=False)
class PortfolioBenchmarks(Resource):
    @limiter.limit(os.getenv("API_RATE_LIMIT", "10 per second"))
    def get(self):
        """List index symbols usable as a benchmark, from the instrument master."""
        api_key = request.args.get("apikey")
        if not api_key or verify_api_key(api_key) is None:
            return make_response(
                jsonify({"status": "error", "message": "Invalid openalgo apikey"}), 403
            )

        try:
            _, payload, status = list_benchmarks()
        except Exception:  # noqa: BLE001
            logger.exception("portfolio benchmarks endpoint failed")
            return make_response(
                jsonify({"status": "error", "message": "Could not list benchmarks."}), 500
            )

        return make_response(jsonify(payload), status)


@api.route("/backtest", strict_slashes=False)
class PortfolioBacktest(Resource):
    @limiter.limit(API_RATE_LIMIT)
    def post(self):
        """Backtest a weighted portfolio and return its full analysis."""
        try:
            data = backtest_schema.load(request.json or {})
        except ValidationError as err:
            return make_response(
                jsonify({"status": "error", "message": err.messages}), 400
            )

        api_key = data.pop("apikey")

        # The key is always verified. Only the *broker session* is optional:
        # a Historify run must not require one, since backtesting at the
        # weekend is the normal case -- but "no broker session needed" is not
        # the same as "no authentication needed", and skipping the check here
        # would have left local history readable by anyone who could reach the
        # endpoint.
        if verify_api_key(api_key) is None:
            return make_response(
                jsonify({"status": "error", "message": "Invalid openalgo apikey"}), 403
            )

        auth_token = feed_token = broker = None
        if data["source"] == "api":
            auth_token, feed_token, broker = get_auth_token_broker(
                api_key, include_feed_token=True
            )
            if auth_token is None:
                return make_response(
                    jsonify(
                        {
                            "status": "error",
                            "message": "No broker session for source='api'. "
                            "Log in to your broker, or use source='db' to "
                            "backtest from local history.",
                        }
                    ),
                    403,
                )

        try:
            success, payload, status = run_portfolio_backtest(
                holdings=data["holdings"],
                start_date=data["start_date"],
                end_date=data["end_date"],
                benchmark=data["benchmark"],
                benchmark_exchange=data["benchmark_exchange"],
                rebalance=data["rebalance"],
                drift_band=data["drift_band"],
                cost_model=data["cost_model"],
                brokerage_pct=data["brokerage_pct"],
                cost_exchange=data["cost_exchange"],
                charge_overrides=data["charges"],
                gst_rate=data["gst_rate"],
                cost_bps=data["cost_bps"],
                slippage=data["slippage"],
                initial_capital=data["initial_capital"],
                risk_free_rate=data["risk_free_rate"],
                source=data["source"],
                api_key=api_key,
                auth_token=auth_token,
                feed_token=feed_token,
                broker=broker,
            )
        except Exception:  # noqa: BLE001
            logger.exception("portfolio backtest endpoint failed")
            return make_response(
                jsonify({"status": "error", "message": "Backtest failed."}), 500
            )

        return make_response(jsonify(payload), status)


@api.route("/tearsheet", strict_slashes=False)
class PortfolioTearsheet(Resource):
    # Heavier than a backtest: openstatz renders a megabyte of embedded charts
    # with matplotlib, which takes a few seconds.
    @limiter.limit(os.getenv("PORTFOLIO_TEARSHEET_RATE_LIMIT", "5 per minute"))
    def post(self):
        """Render the full openstatz tearsheet as a downloadable HTML file."""
        try:
            data = backtest_schema.load(request.json or {})
        except ValidationError as err:
            return make_response(
                jsonify({"status": "error", "message": err.messages}), 400
            )

        api_key = data.pop("apikey")
        if verify_api_key(api_key) is None:
            return make_response(
                jsonify({"status": "error", "message": "Invalid openalgo apikey"}), 403
            )

        auth_token = feed_token = broker = None
        if data["source"] == "api":
            auth_token, feed_token, broker = get_auth_token_broker(
                api_key, include_feed_token=True
            )
            if auth_token is None:
                return make_response(
                    jsonify(
                        {
                            "status": "error",
                            "message": "No broker session for source='api'.",
                        }
                    ),
                    403,
                )

        try:
            ok, payload, status = generate_tearsheet(
                holdings=data["holdings"],
                start_date=data["start_date"],
                end_date=data["end_date"],
                benchmark=data["benchmark"],
                benchmark_exchange=data["benchmark_exchange"],
                rebalance=data["rebalance"],
                drift_band=data["drift_band"],
                cost_model=data["cost_model"],
                brokerage_pct=data["brokerage_pct"],
                cost_exchange=data["cost_exchange"],
                charge_overrides=data["charges"],
                gst_rate=data["gst_rate"],
                cost_bps=data["cost_bps"],
                slippage=data["slippage"],
                initial_capital=data["initial_capital"],
                risk_free_rate=data["risk_free_rate"],
                source=data["source"],
                api_key=api_key,
                auth_token=auth_token,
                feed_token=feed_token,
                broker=broker,
            )
        except Exception:  # noqa: BLE001
            logger.exception("portfolio tearsheet endpoint failed")
            return make_response(
                jsonify(
                    {
                        "status": "error",
                        "message": "Tearsheet generation failed.",
                    }
                ),
                500,
            )
        if not ok:
            return make_response(jsonify(payload), status)

        # Served as an attachment: it is a self-contained report a user keeps,
        # not a page the app renders.
        response = make_response(payload, 200)
        response.headers["Content-Type"] = "text/html; charset=utf-8"
        response.headers["Content-Disposition"] = (
            'attachment; filename="portfolio-tearsheet.html"'
        )
        return response


class HoldingsAnalysisSchema(Schema):
    apikey = fields.Str(required=True, validate=validate.Length(min=1, max=256))
    # How much history to judge the current holdings against. A year is the
    # shortest window that contains a full seasonal cycle.
    lookback_days = fields.Int(load_default=365, validate=validate.Range(min=60, max=3650))
    benchmark = fields.Str(load_default="NIFTY", allow_none=True)
    benchmark_exchange = fields.Str(
        load_default="NSE_INDEX", validate=validate.OneOf(BENCHMARK_EXCHANGES)
    )
    risk_free_rate = fields.Float(load_default=0.0, validate=validate.Range(min=0, max=0.5))
    source = fields.Str(load_default="db", validate=validate.OneOf(["db", "api"]))


holdings_schema = HoldingsAnalysisSchema()


@api.route("/holdings", strict_slashes=False)
class PortfolioHoldings(Resource):
    @limiter.limit(API_RATE_LIMIT)
    def post(self):
        """Analyse the portfolio actually held at the broker."""
        try:
            data = holdings_schema.load(request.json or {})
        except ValidationError as err:
            return make_response(
                jsonify({"status": "error", "message": err.messages}), 400
            )

        api_key = data.pop("apikey")
        if verify_api_key(api_key) is None:
            return make_response(
                jsonify({"status": "error", "message": "Invalid openalgo apikey"}), 403
            )

        # Holdings always need a broker session -- unlike a backtest, there is
        # no local copy of what someone owns.
        auth_token, feed_token, broker = get_auth_token_broker(
            api_key, include_feed_token=True
        )
        if auth_token is None:
            return make_response(
                jsonify(
                    {
                        "status": "error",
                        "message": "No broker session. Log in to your broker to "
                        "analyse live holdings.",
                    }
                ),
                403,
            )

        try:
            _, payload, status = analyse_live_holdings(
                lookback_days=data["lookback_days"],
                benchmark=data["benchmark"],
                benchmark_exchange=data["benchmark_exchange"],
                risk_free_rate=data["risk_free_rate"],
                source=data["source"],
                api_key=api_key,
                auth_token=auth_token,
                feed_token=feed_token,
                broker=broker,
            )
        except Exception:  # noqa: BLE001
            logger.exception("holdings analysis failed")
            return make_response(
                jsonify({"status": "error", "message": "Analysis failed."}), 500
            )

        return make_response(jsonify(payload), status)
